Add -fhardened to Compiler Options Hardening Guide

thomasnyman opened the pull request:
Adds [`-fhardened](https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html#index-fhardened) introduced in [GCC 14](https://gcc.gnu.org/gcc-14/changes.html) to Compiler Options Hardening Guide.

Fixes #240

thomasnyman commented:
Commit 01973987f6b6782faeb4cf63beac8f785f9ae217 fixes version information to be consistent with #507.

siddhesh commented on docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md:
```diff
@@ -987,6 +988,34 @@ The `-fexceptions` option is also needed for C code that needs to interoperate w
+
+### Additional Considerations
+
+Options explicitly specified on the compiler command line always take precedence, e.g., `-fhardened -fstack-protector` will enable `-fstack-protector`, not `-fstack-protector-strong` implied by `-fhardened`.
```
There's `-Whardened`, which should warn if a flag ends up downgrading or disabling hardening enabled by `-fhardened`. It should get enabled whenever `-fhardened` is specified.

thomasnyman replied:
Thanks, that's a very useful addition. I'll add it to the table in this section as well.

thomasnyman commented:
Commit fbc1614eaf862bc872cee9d8f4eebb811953e966 addresses the feedback from the [C/C++ Compiler BP Guide 2024-05-30](https://docs.google.com/document/d/1JY8FREBPCUUFpuv7-4B9EjeS2MLDpel0dbG5DFWrTns/edit#bookmark=id.9h6u9sw0ldw).
